### PR TITLE
Add calctextsize overload and remove InputTextWithHint from generation

### PR DESCRIPTION
**Pull request by zaafar**
This MR fixes https://github.com/mellinoe/ImGui.NET/issues/216

Manual Testing Done
```
                // Font size 13.
                string textwith1Char = "a";
                string textwith2Char = "aa";
                string textwith2Hashes = "aaaa##aaaa";
                ImGui.Text($"TC1: 0  Char Expected output: {ImGui.CalcTextSize(null).X}"); // actual -> 0 / 7 = 0 char
                ImGui.Text($"TC2: 1  Char Expected output: {ImGui.CalcTextSize(textwith1Char).X}"); // actual -> 7 / 7 = 1 char
                ImGui.Text($"TC3: 2  Char Expected output: {ImGui.CalcTextSize(textwith2Char).X}"); // actual -> 14 / 7 = 2 char
                ImGui.Text($"TC4: 1  Char Expected output: {ImGui.CalcTextSize(textwith2Char, 0, 1).X}"); // actual -> 7/ 7 = 1 char
                ImGui.Text($"TC5: 1  Char Expected output: {ImGui.CalcTextSize(textwith2Char, 1, 1).X}"); // actual -> 7/ 7 = 1 char
                ImGui.Text($"TC6: 0  Char Expected output: {ImGui.CalcTextSize(textwith2Char, 1, 0).X}"); // actual -> 0 / 7 = 0 char
                ImGui.Text($"TC7: 0  Char Expected output: {ImGui.CalcTextSize(textwith2Char, 0, 0).X}"); // actual -> 0 / 7 = 0 char
                ImGui.Text($"TC8: 4  Char Expected output: {ImGui.CalcTextSize(textwith2Hashes, hideTextAfterDoubleHash: true).X}"); // actual -> 28 / 7 = 4 char
                ImGui.Text($"TC9: 10 Char Expected output: {ImGui.CalcTextSize(textwith2Hashes, hideTextAfterDoubleHash: false).X}"); // actual -> 70 / 7 = 10 char
                ImGui.Text($"TCA: 4  Char Expected output: {ImGui.CalcTextSize(textwith2Hashes, wrapWidth: 4 * 7f).X}"); // actual -> 28 / 7 = 4 char
                ImGui.Text($"TCB: 3 lines Expected output: {ImGui.CalcTextSize(textwith2Hashes, wrapWidth: 4 * 7f).Y}"); // actual -> 39 / 13 = 3 lines



                // Font size 13.
                string textwith1Char = "あ";
                string textwith2Char = "ああ";
                string textwith2Hashes = "ああああ##ああああ";
                ImGui.Text($"TC1: 0  Char Expected output: {ImGui.CalcTextSize(null).X}"); // actual -> 0 / 7 = 0 char
                ImGui.Text($"TC2: 1  Char Expected output: {ImGui.CalcTextSize(textwith1Char).X}"); // actual -> 7 / 7 = 1 char
                ImGui.Text($"TC3: 2  Char Expected output: {ImGui.CalcTextSize(textwith2Char).X}"); // actual -> 14 / 7 = 2 char
                ImGui.Text($"TC4: 1  Char Expected output: {ImGui.CalcTextSize(textwith2Char, 0, 1).X}"); // actual -> 7 / 7 = 1 char
                ImGui.Text($"TC5: 1  Char Expected output: {ImGui.CalcTextSize(textwith2Char, 1, 1).X}"); // actual -> 7 / 7 = 1 char
                ImGui.Text($"TC6: 0  Char Expected output: {ImGui.CalcTextSize(textwith2Char, 1, 0).X}"); // actual -> 0 / 7 = 0 char
                ImGui.Text($"TC7: 0  Char Expected output: {ImGui.CalcTextSize(textwith2Char, 0, 0).X}"); // actual -> 0 / 7 = 0 char
                ImGui.Text($"TC8: 4  Char Expected output: {ImGui.CalcTextSize(textwith2Hashes, hideTextAfterDoubleHash: true).X}"); // actual -> 28 / 7 = 4 char
                ImGui.Text($"TC9: 10 Char Expected output: {ImGui.CalcTextSize(textwith2Hashes, hideTextAfterDoubleHash: false).X}"); // actual -> 70 / 7 = 10 char
                ImGui.Text($"TCA: 4  Char Expected output: {ImGui.CalcTextSize(textwith2Hashes, wrapWidth: 4 * 7f).X}"); // actual -> 28 / 7 = 4 char
                ImGui.Text($"TCB: 3 lines Expected output: {ImGui.CalcTextSize(textwith2Hashes, wrapWidth: 4 * 7f).Y}"); // actual -> 39 / 13 = 3 lines
```

### CODE REVIEW

**Comment by zaafar** [comment]
Let me know if ppl still need `igInputTextWithHint` generated function and I will revert that part of the MR.

**Comment by sonoro1234** [comment]
Why not just use igCalcTextSize with second argument NULL and let imgui do the work?

**Comment by zaafar** [comment]
@sonoro1234 how will that work for TC-4,5,6,7 which are valid use-cases.

**Comment by sonoro1234** [comment]
> TC-4,5,6,7?

Sorry, I dont know what you mean with `TC-4,5,6,7?`

**Comment by sonoro1234** [comment]
I am not sure to understand.
I mean than in
````
ImGuiNative.igCalcTextSize(&ret, nativeTextStart, nativeTextEnd, *((byte*)(&hideTextAfterDoubleHash)), wrapWidth);
````
you can set nativeTextEnd to null and ImGuiNative.igCalcTextSize is supposed to do the same work.

**Comment by zaafar** [comment]
Sorry, let me explain in details. The drawback of asking the caller to use `ImGuiNative.igCalcTextSize(&ret, nativeTextStart, nativeTextEnd, *((byte*)(&hideTextAfterDoubleHash)), wrapWidth);` is as following:

1: Caller can not directly use c# strings. They have to convert string to byte* (i.e. more burden on the caller).

2: Caller can not CalcTextSize on half a string or X chars of a string without doing any string manipulation. So more burden on the caller if they want to do something like `ImGui.CalcTextSize("Hello World", length: "Hello".Length)`

So basically adding this helper function in ImGui.NET lib ensures that caller doesn't have to worry about all these things.

File: src/ImGui.NET/ImGui.Manual.cs [hunk 1]
```diff
@@ -355,6 +355,47 @@ public static unsafe partial class ImGui
...
+            Vector2 ret;
+            byte* nativeText = null;
+            byte* nativeTextStart = null;
+            byte* nativeTextEnd = null;
+            int textByteCount = 0;
+            int textByteSize = 0;
+            if (text != null)
+            {
+                textByteCount = Encoding.UTF8.GetByteCount(text);
+                textByteSize = Encoding.UTF8.GetByteCount("X");
```
Comment (mellinoe) [inline src/ImGui.NET/ImGui.Manual.cs:374]: The usage of this variable doesn't seem correct to me. "X" will always have a 1-byte UTF8 representation, so computing it here seems unnecessary. But not all C# chars will have a 1-byte representation, particularly in other languages. I'm not sure exactly how to address this, presumably you would need to scan through the string to find the offset of the particular element you need.

File: src/ImGui.NET/Generated/ImGui.gen.cs [hunk 2]
```diff
@@ -8085,297 +8054,6 @@ public static bool InputScalarN(string label, ImGuiDataType data_type, IntPtr p_
             }
             return ret != 0;
         }
-        public static bool InputTextWithHint(string label, string hint, string buf, uint buf_size)
```
Comment (mellinoe) [inline src/ImGui.NET/Generated/ImGui.gen.cs:8088]: Why was this removed?

**Review by mellinoe — CHANGES_REQUESTED** [review]

File: src/ImGui.NET/Generated/ImGui.gen.cs (hunk 2 above)
Comment (zaafar) [reply src/ImGui.NET/Generated/ImGui.gen.cs:8088]: Issue https://github.com/mellinoe/ImGui.NET/issues/215 shows that the **ImGui.gen** default implementation of the `InputTextWithHint`  feature is basically useless. This commit (https://github.com/mellinoe/ImGui.NET/commit/66f1afa9a268c2b70ff9403c570e45c8847e37a4) fixes that issue by manually adding public APIs for `InputTextWithHint`  feature in **ImGui.Manual.cs** class.

With that being said, in my opinion, in order to not confuse the users of this lib, it would be nice to remove the default implementation of `InputTextWithHint`  feature from  **ImGui.gen** class. In order to do that, I added this line (https://github.com/mellinoe/ImGui.NET/pull/221/files#diff-c5973b39c150ae89c1170e041fddade4da7998b0d994fad05394f4b311e402a9R104) in this MR and run the generator tool.

Let me know if you disagree with this thinking process and I will revert this part of the MR.

File: src/ImGui.NET/ImGui.Manual.cs (hunk 1 above)
Comment (zaafar) [reply src/ImGui.NET/ImGui.Manual.cs:374]: Agree, this isn't perfect/currect.

->  you would need to scan through the string to find the offset of the particular element you need.

assuming `you` means caller over here, rather than accepting `start` and `length`, the function can accept `startInByteSize` and `lengthInByteSize` to put the burden on the caller to figure out the `startOffsetInByte` and `LengthOffsetInByte`.

File: src/ImGui.NET/Generated/ImGui.gen.cs (hunk 2 above)
Comment (mellinoe) [reply src/ImGui.NET/Generated/ImGui.gen.cs:8088]: I forgot that the manual overloads had already been merged in for TextWithHint. That clear it up, thanks.

File: src/ImGui.NET/ImGui.Manual.cs (hunk 1 above)
Comment (mellinoe) [reply src/ImGui.NET/ImGui.Manual.cs:374]: I think the current signature of this function is correct, with `start` and `length` referring to C# char indices. This is what 99.9% of users will expect, because all other string-related functions using "start" and "length" behave that way, so I don't think we should push anything onto the caller. However, when converting to UTF8 and UTF8-based offsets, we might need to do something different with the computed addresses, because they aren't necessarily 1-to-1 with char indices.

Comment (zaafar) [reply src/ImGui.NET/ImGui.Manual.cs:374]: Looks like it's not an easy problem to solve. I am open to suggestions/code-changes. Following is the best I can come up with.

https://gist.github.com/zaafar/ccf2c32a9fa4d85cbe0dcd7163beb81d#file-imgui-manual-cs-L46

NOTE: The code is in public gist because I don't want to pollute the PR with potential solution. Sorry about that.

Comment (zaafar) [reply src/ImGui.NET/ImGui.Manual.cs:374]: Drawback (that I can think of): ReadOnlySpan is not available in .net standard 2.0.

Comment (zaafar) [reply src/ImGui.NET/ImGui.Manual.cs:374]: other than using SubString(start, length) or AsSpan(start, length), I couldn't figure out any way to convert `C# char indices` to `UTF8 based offsets`.

Comment (zaafar) [reply src/ImGui.NET/ImGui.Manual.cs:374]: 1 benefit is that, only the portion of the string that needs to be computed by `CalcTextSize` is allocated via `Util.Allocate`/`Util.Allocate` function.

File: src/ImGui.NET/ImGui.Manual.cs [hunk 3]
```diff
@@ -355,6 +355,47 @@ public static unsafe partial class ImGui
             return result != 0;
         }
 
+        public static Vector2 CalcTextSize(
+            string text,
+            int? start = null,
```
[the remark this answers is not in the document]
Comment (zaafar) [reply src/ImGui.NET/ImGui.Manual.cs]: done. let me know if u need anything else.

File: src/ImGui.NET/ImGui.Manual.cs (hunk 1 above)
Comment (zaafar) [reply src/ImGui.NET/ImGui.Manual.cs]: never mind, found a solution that is better and works. so done.

**Comment by zaafar** [comment]
@mellinoe let me know if you need any further changes.
